src/junit.rs

```diff
@@ -65,104 +66,173 @@ fn populate(rel_attrs: RelevantAttrs, testsuite: String) -> PyResult<Testrun> {
+                    if e.name().as_ref() == b"testcase" {
+                        let rel_attrs = get_relevant_attrs(e.attributes())?;
+                        let testrun = populate(rel_attrs, curr_testsuite.clone())?;
+                        on_testcase_finish(testrun, curr_testsuite.clone());
+                    }
                 }
-                _ => {}
-            },
-            Ok(Event::End(e)) => match e.name().as_ref() {
-                b"testcase" => {
-                    list_of_test_runs.push(
-                        saved_testrun
-                            .ok_or(ParserError::new_err("Error accessing saved testrun"))?,
-                    );
-                    saved_testrun = None;
+                Ok(Event::Text(x)) => {
+                    if in_failure {
+                        let testrun = saved_testrun
+                            .as_mut()
+                            .ok_or(ParserError::new_err("Error accessing saved testrun"))?;
+
+                        let mut xml_failure_message = x.into_owned();
+                        xml_failure_message.inplace_trim_end();
+                        xml_failure_message.inplace_trim_start();
+
+                        testrun.failure_message =
+                            Some(String::from_utf8(xml_failure_message.as_ref().to_vec())?);
+                    }
                 }
-                b"failure" => in_failure = false,
+
+                // There are several other `Event`s we do not consider here
                 _ => (),
-            },
-            Ok(Event::Empty(e)) => {
-                if e.name().as_ref() == b"testcase" {
-                    let rel_attrs = get_relevant_attrs(e.attributes())?;
-                    list_of_test_runs.push(populate(rel_attrs, curr_testsuite.clone())?);
-                }
-            }
-            Ok(Event::Text(x)) => {
-                if in_failure {
-                    let testrun = saved_testrun
-                        .as_mut()
-                        .ok_or(ParserError::new_err("Error accessing saved testrun"))?;
-
-                    let mut xml_failure_message = x.into_owned();
-                    xml_failure_message.inplace_trim_end();
-                    xml_failure_message.inplace_trim_start();
-
-                    testrun.failure_message =
-                        Some(String::from_utf8(xml_failure_message.as_ref().to_vec())?);
-                }
             }
-
-            // There are several other `Event`s we do not consider here
-            _ => (),
+            buf.clear()
+        };
+
+    match loop_result {
+        Ok(testruns) => {
+            // detect framework based on all the information we'e collected so far
+            let framework = detect_framework(
+                testsuites_name,
+                testsuite_names,
+                filenames,
+                example_class_name,
+                example_test_name,
+                failure_messages,
+            );
```

Comment:
this is interesting, as you are collecting all of this information above just for this single call. And all this information is duplicated with the information thats in `Testrun`, except for `testsuites_name`.

I would maybe add a `Testrun::framework()` or `Testrun::detect_framework(&self) -> Option<Framework>` function. Then you can iterate over the testruns, and break out when the first testrun yields a possible framework.

For `testsuites_name`, you can also just keep a `let mut detected_framework = None` around, and write to that directly if it `is_none()`, and you have a `<testsuites name="...">` available.